{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4928ce67",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "from Bio import Entrez\n",
    "import os\n",
    "import time"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5c485ff0",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pd.read_csv(\"ecoli_only_100000.csv\")\n",
    "# Konfiguracja Entrez\n",
    "Entrez.email = email\n",
    "# Upewnij się, że Twój klucz API jest poprawny\n",
    "Entrez.api_key = apikey\n",
    "\n",
    "# Lista Twoich numerów akcesyjnych\n",
    "accessions = list(df.accession.unique())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c618d52c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Folder na wyniki\n",
    "output_dir = \"ecoli_genomes\"\n",
    "if not os.path.exists(output_dir):\n",
    "    os.makedirs(output_dir)\n",
    "\n",
    "\n",
    "def download_sequences(acc_list):\n",
    "    print(f\"Rozpoczynam pobieranie {len(acc_list)} sekwencji...\")\n",
    "\n",
    "    for acc in acc_list:\n",
    "        file_path = os.path.join(output_dir, f\"data/complete_ecoli/{acc}.fasta\")\n",
    "\n",
    "        # Pomijamy, jeśli plik już istnieje\n",
    "        if os.path.exists(file_path):\n",
    "            print(f\"[*] Plik {acc}.fasta już istnieje, pomijam.\")\n",
    "            continue\n",
    "\n",
    "        try:\n",
    "            print(f\"[>] Pobieranie {acc}...\")\n",
    "            # efetch pobiera dane bezpośrednio z bazy\n",
    "            handle = Entrez.efetch(\n",
    "                db=\"nucleotide\", id=acc, rettype=\"fasta\", retmode=\"text\"\n",
    "            )\n",
    "            fasta_data = handle.read()\n",
    "            handle.close()\n",
    "\n",
    "            # Zapis do pliku\n",
    "            with open(file_path, \"w\") as f:\n",
    "                f.write(fasta_data)\n",
    "\n",
    "            print(f\"[OK] Zapisano w {file_path}\")\n",
    "\n",
    "            # Przerwa dla serwerów NCBI (max 3 zapytania na sekundę bez API Key)\n",
    "            time.sleep(0.5)\n",
    "\n",
    "        except Exception as e:\n",
    "            print(f\"[BŁĄD] Nie udało się pobrać {acc}: {e}\")\n",
    "\n",
    "\n",
    "download_sequences(accessions)\n",
    "print(\"\\nProces zakończony.\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "base",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.13.11"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
